### PR TITLE
json rpc: use checkpoint timestamp for txn response instead

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

Today txn timestamp is a local FN timestamp, which is not consistent across FNs, now as we have timestamp in checkpoint, this PR uses that as the consistent txn timestamp.

## Test Plan 
read a checkpoint locally, verify that the timestamp of txns in the checkpoint matches the checkpoint timestamp

```
curl --location --request POST http://127.0.0.1:9000 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getCheckpoint", "params":[20]}'
{"jsonrpc":"2.0","result":{"epoch":0,"sequenceNumber":20,"digest":"4TyV6KjrdPUVVyF1GV2P8tsg4QFdnMDjK7HdKWbJruT6","networkTotalTransactions":4,"previousDigest":"95NLNPr8v1gEtME35rgPzCchnviz77oXkpaw5kUSPqNe","epochRollingGasCostSummary":{"computation_cost":435,"storage_cost":144,"storage_rebate":64},"timestampMs":1677254340992,"endOfEpochData":null,"transactions":["GRM8UHr47nLdi1HgrYkGccQv7djtk2F73gLV6HZTxXRN","8tjEtPY8KpP6tpwpyA1YhHk1qW5SBcaLweqfdAgqvage"]},"id":1}%        
                                                                                    
curl --location --request POST http://127.0.0.1:9000 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getTransaction", "params":["GRM8UHr47nLdi1HgrYkGccQv7djtk2F73gLV6HZTxXRN"]}'
... timestampMs":1677254340992,"checkpoint":20},"id":1}%                                                                                                                                                                         

curl --location --request POST http://127.0.0.1:9000 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getTransaction", "params":["8tjEtPY8KpP6tpwpyA1YhHk1qW5SBcaLweqfdAgqvage"]}'
..."timestampMs":1677254340992,"checkpoint":20},"id":1}%        
```

---
### Type of Change (Check all that apply)

- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-json-rpc/src/read_api.rs [hunk 1]
```diff
@@ -152,17 +152,27 @@ impl ReadApiServer for ReadApi {
...
+            .map_err(|e| anyhow!("{e}"))?
+            .map(|(_epoch, seq)| seq);
+        let mut checkpoint = None;
+        if let Some(checkpoint_seq) = checkpoint_seq_opt {
+            checkpoint = Some(
+                self.get_checkpoint(CheckpointId::SequenceNumber(checkpoint_seq))
+                    .await?,
+            );
+        }
+        let checkpoint_timestamp = checkpoint.map(|c| c.timestamp_ms);
```
Comment (lxfind) [inline crates/sui-json-rpc/src/read_api.rs:168]: Can all be done in a one sequence of maps. Specifically, you would want to not error out even if there is an error when reading checkpoint (just make the timestamp None), which I think is preferred.

File: crates/sui-json-rpc/src/read_api.rs [hunk 2]
```diff
@@ -152,17 +152,29 @@ impl ReadApiServer for ReadApi {
...
-        let checkpoint = self
+        let checkpoint_seq_opt = self
             .state
             .database
             .get_transaction_checkpoint(&digest)
-            .map_err(|e| anyhow!("{e}"))?;
+            .map_err(|e| anyhow!("{e}"))?
+            .map(|(_epoch, seq)| seq);
+        let mut checkpoint_timestamp = None;
+        if let Some(checkpoint_seq) = checkpoint_seq_opt {
```
Comment (lxfind) [inline crates/sui-json-rpc/src/read_api.rs:162]: You can assign to checkpoint_timestamp directly from map

File: crates/sui-json-rpc/src/read_api.rs (hunk 1 above)
Comment (gegaowp) [reply crates/sui-json-rpc/src/read_api.rs]: thanks, made a change while I firstly tried map with an async closure like below and saw an `async closures are unstable` error, any wisdom?
```
        let checkpoint = checkpoint_seq_opt
            .map(async move |seq| {
                let tmp = self.get_checkpoint(CheckpointId::SequenceNumber(seq))
                    .await?;
                tmp
            });
```

**Review by lxfind — APPROVED** [review]

File: crates/sui-json-rpc/src/read_api.rs (hunk 2 above)
Comment (gegaowp) [reply crates/sui-json-rpc/src/read_api.rs:162]: yeah I saw the "async closures are unstable error" can I did that, did I miss sth?

Comment (lxfind) [reply crates/sui-json-rpc/src/read_api.rs:162]: Ah I see. Free free to merge then :)

**Pull request merged**